frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,183 @@
+
+import com.virtuslab.gitmachete.backend.api.BaseGitMacheteBranch;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRemoteBranch;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRepository;
+import com.virtuslab.gitmachete.backend.api.SyncToRemoteStatus;
+import com.virtuslab.gitmachete.frontend.actions.contextmenu.CheckoutSelectedBranchAction;
+import com.virtuslab.gitmachete.frontend.datakeys.DataKeys;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * Expects DataKeys:
+ * <ul>
+ *  <li>{@link DataKeys#KEY_GIT_MACHETE_REPOSITORY}</li>
+ *  <li>{@link DataKeys#KEY_SELECTED_VCS_REPOSITORY}</li>
+ *  <li>{@link CommonDataKeys#PROJECT}</li>
+ * </ul>
+ */
+public abstract class BaseResetBranchToRemoteAction extends GitMacheteRepositoryReadyAction implements IBranchNameProvider {
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+  private static final String VCS_NOTIFIER_TITLE = "Resetting";
+  private static final String TASK_TITLE = "Resetting...";
+
+  protected abstract String getActionType();
+
+  @Override
+  @UIEffect
+  public void update(AnActionEvent anActionEvent) {
+    super.update(anActionEvent);
+
+    Presentation presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    var branchName = getNameOfBranchUnderAction(anActionEvent);
+
+    if (branchName.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled due to undefined ${getActionType()} branch");
+      return;
+    }
+
+    String branchNameString = branchName.get();
+
+    Option<SyncToRemoteStatus> syncToRemoteStatus = getGitMacheteRepository(anActionEvent)
+        .flatMap(repo -> repo.getBranchByName(branchNameString))
+        .map(BaseGitMacheteBranch::getSyncToRemoteStatus);
+
+    if (syncToRemoteStatus.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled due to undefined sync to remote status");
+      return;
+    }
+
+    SyncToRemoteStatus.Relation relation = syncToRemoteStatus.get().getRelation();
+
+    if (relation != SyncToRemoteStatus.Relation.Untracked) {
+      presentation.setDescription("Reset branch '${branchNameString} to its remote tracking branch");
+    } else {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled because ${getActionType()} branch is untracked");
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    LOG.debug("Performing");
+
+    Project project = getProject(anActionEvent);
+    var gitRepository = getSelectedVcsRepository(anActionEvent);
+    var branchName = getNameOfBranchUnderAction(anActionEvent);
+    var macheteRepository = getGitMacheteRepository(anActionEvent);
+
+    if (branchName.isDefined()) {
```

Comment:
Maybe flip the ifs `to `if (NOT ...) { LOG...; return; }` to reduce nesting & keep warnings close to their respective conditions